src/browser/application.ts

```diff
@@ -0,0 +1,111 @@
+import { Logger } from "pino";
+import { v4 as uuid } from "uuid";
+import { IApplication } from "../base/application";
+import { IInputHandler, IInputMessage } from "../base/input-handler";
+import { ISignalProvider } from "../base/signal-provider";
+import { IStreamProvider } from "../base/stream-provider";
+import { IWebrtcProvider } from "../base/webrtc-provider";
+
+/**
+ * Application constructor options
+ */
+export interface IApplicationOpts {
+  /**
+   * A logger
+   */
+  logger: Logger;
+
+  /**
+   * The name of the capture window
+   */
+  captureWindowTitle: string;
+
+  /**
+   * The ice servers to use
+   */
+  iceServers: RTCIceServer[];
+
+  /**
+   * The signal provider to use
+   */
+  signalProvider: ISignalProvider;
+
+  /**
+   * The stream provider to use
+   */
+  streamProvider: IStreamProvider;
+
+  /**
+   * The webrtc provider to use
+   */
+  webrtcProvider: IWebrtcProvider;
+
+  /**
+   * The input handler to use
+   */
+  inputHandler: IInputHandler;
+}
+
+/**
+ * A browser application - orchestrates the streamer experience
+ */
+export class Application implements IApplication {
+  private opts: IApplicationOpts;
+
+  /**
+   * Default ctor
+   * @param opts ctor options
+   */
+  constructor(opts: IApplicationOpts) {
+    this.opts = opts;
+  }
+
+  /**
+   * Internal boot helper
+   */
+  public async boot() {
+    const {
+      logger,
+      captureWindowTitle,
+      iceServers,
+      signalProvider,
+      streamProvider,
+      webrtcProvider,
+      inputHandler } = this.opts;
+    logger.info("Browser: initializing application");
+
+    const devices = await streamProvider.enumerateDevices((e) => e.name === captureWindowTitle);
+    const selectedDevice = devices[0];
```

Comment:
We assume that we can always find the selected device. Should we add some error condition checking here?

Reply:
we can fail with a more explicit error, sure that seems like a good idea. We can't actually recover or anything though.

Reply:
Yeah, agree.

Reply:
this is also the source of [the above failure](https://github.com/bengreenier/browserd/pull/23#issuecomment-507302980), so i'll report status on it here.

Reply:
added more detailed logging, failure cases. also a test to validate no device error case.